File: src/main/java/hu/elte/wr14yr/musicportal/service/SearchService.java
package hu.elte.wr14yr.musicportal.service;

import hu.elte.wr14yr.musicportal.model.*;
import hu.elte.wr14yr.musicportal.model.tags.*;
import hu.elte.wr14yr.musicportal.model.tags.PlaylistTag;
import hu.elte.wr14yr.musicportal.repository.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.logging.Logger;

@Service
public class SearchService {
    @Autowired
    private SongRepository songRepository;

    @Autowired
    private AlbumRepository albumRepository;

    @Autowired
    private PlaylistRepository playlistRepository;

    @Autowired
    private UserRepository userRepository;

    private Logger logger = Logger.getLogger(SearchService.class.getName());

    public Iterable<Album> findAlbumsByTitle(String searchWord) {
        logger.info(String.format("Search service: albums with title %s" +
                " are going to be searched", searchWord));

        return albumRepository.findAllByTitleContainsAllIgnoreCase(searchWord);
    }

    public Iterable<Album> findAlbumByAlbumTag(String tagWord) {
        logger.info("Search service: albums with current album tag are going to be searched");

        return albumRepository.findAllByAlbumTagsWordContainsAllIgnoreCase(tagWord);
    }

    public Iterable<Album> findAlbumsByGenre(String genreWord) {
        logger.info("Search service: albums with current genres are going to be searched");

        return albumRepository.findAllByGenresNameEquals(genreWord);
    }

    public Iterable<Song> findSongsByTitle(String searchWord) {
        logger.info(String.format("Search service: songs with title %s are going to be searched", searchWord));

        return songRepository.findAllByTitleContainsAllIgnoreCase(searchWord);
    }

    public Iterable<Playlist> findPlaylistByName(String searchWord) {
        logger.info(String.format("Search service: playlists with name %s are going to be searched", searchWord));

        return playlistRepository.findAllByNameContainsAllIgnoreCase(searchWord);
    }
    public Iterable<Playlist> findPlaylistByPlaylistTag(String tagWord) {
        logger.info("Search service: playlists with current playlist tag are going to be searched");

        return playlistRepository.findAllByPlaylistTagsWordContainsAllIgnoreCase(tagWord);
    }

    public Iterable<User> findUsersByUserTag(String tagWord) {
        logger.info("Search service: users with current user tag are going to be searched");

        return userRepository.findAllByUserTagsWordContainsAllIgnoreCase(tagWord);
    }

    public Iterable<User> findUsersByGenre(String genreWord) {
        logger.info("Search service: users with current favourite genre are going to be searched");

        return userRepository.findAllByFavGenreIdNameEquals(genreWord);
    }
}

File: src/main/frontend/src/app/service/search.service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Album} from "../model/album";
import {AlbumTag} from "../model/tags/albumtag";
import {Genre} from "../model/genre";
import {User} from "../model/user";
import {Playlist} from "../model/playlist";
import {PlaylistTag} from "../model/tags/playlisttag";
import {UserTag} from "../model/tags/usertag";
import {Tag} from "../model/tags/tag";
import {Song} from "../model/song";

@Injectable({
  providedIn: 'root'
})
export class SearchService {

  constructor(
    private http: HttpClient
  ) { }

  getAlbumsByTitle(searchWord: string): Promise<Album[]> {
    return this.http.get<Album[]>(
      `api/search/albums/${searchWord}`
    ).toPromise();
  }

  getAlbumsByAlbumTag(tagWord: string): Promise<Album[]> {
    const formData = new FormData();
    formData.append("tag", tagWord);
    return this.http.post<Album[]>(
      "api/search/albums/tag",
      formData
    ).toPromise();
  }

  getAlbumsByGenre(genreWord: string): Promise<Album[]> {
    const formData = new FormData();
    formData.append("genre", genreWord);
    return this.http.post<Album[]>(
      "api/search/albums/genre",
      formData
    ).toPromise();
  }

  getSongsByTitle(searchWord: string): Promise<Song[]> {
    return this.http.get<Song[]>(
      `api/search/songs/${searchWord}`
    ).toPromise();
  }

  getPlaylistByName(searchWord: string): Promise<Playlist[]> {
    return this.http.get<Playlist[]>(
      `api/search/playlists/${searchWord}`
    ).toPromise();
  }

  getPlaylistByPlaylistTag(tagWord: string): Promise<Playlist[]> {
    const formData = new FormData();
    formData.append("tag", tagWord);
    return this.http.post<Playlist[]>(
      "api/search/playlists/tag",
      formData
    ).toPromise();
  }

  getUsersByUserTag(tagWord: string): Promise<User[]> {
    const formData = new FormData();
    formData.append("tag", tagWord);
    return this.http.post<User[]>(
      "api/search/users/tag",
      formData
    ).toPromise();
  }

  getUsersByGenre(genreWord: string): Promise<User[]> {
    const formData = new FormData();
    formData.append("genre", genreWord);
    return this.http.post<User[]>(
      "api/search/users/genre",
      formData
    ).toPromise();
  }
}
